docs/subnets/introduction-to-vm.md

```diff
@@ -0,0 +1,335 @@
+
+### Handlers
+
+Handlers serve the response for the incoming HTTP requests. Handlers can also be wrapped with **gRPC** for efficiently making calls from other services such as `AvalancheGo`. Handlers help in creating APIs. VM implements 2 kinds of handlers:
+
+- **Non-Static Handlers** - They help in interacting with blockchains instantiated by the VM. The API's endpoint will be different for different chains. `/ext/bc/[chainID]`
+- **Static Handlers** - They help in directly accessing VM. These are optional for reasons such as parsing genesis bytes required to instantiate new blockchains. `/ext/vm/[vmID]`
+
+For any developers familiar with object-oriented programming, this is very similar to static and non-static methods on a class.
+
+### VM Factory
+
+Each VM has a **factory** that is capable of creating new VM instances. A VM can be initialized as a blockchain along with handlers for accessing it, only when we have the VM's instance. The factory's `New` method shown below, returns the VM's instance to its caller in `AvalancheGo`. It's generally in the [`factory.go`](https://github.com/ava-labs/blobvm/blob/master/factory.go) file of the VM.
+
+```go
+// Returning a new VM instance from VM's factory
+func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
+```
+
+### Initializing a VM
+
+Blockchains are functional, only when the instantiated VMs are initialized and the node is bootstrapped. Initializing a VM involves setting up the database, block builder, mempool, genesis state, handlers, etc. This will expose the VM's API handlers, start accepting transactions, gossiping them across the network, building blocks, etc. More details on it can be found in the [third part](/subnets/create-a-vm-blobvm#initialize) of this series.
+
+```go
+if err := vm.Initialize(
+    ctx.Context,
+    vmDBManager,
+    genesisData,
+    chainConfig.Upgrade,
+    chainConfig.Config,
+    msgChan,
+    fxs,
+    sender,
+);
+```
+
+You can refer to the [implementation](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L92) of `vm.initialize` in the BlobVM repository.
+
+### Interaction
+
+We can use the API handlers to issue transactions, query chain state, and other functionalities provided by the VM.
+
+## Interfaces
+
+The main task of a VM is to issue blocks to the consensus engine whenever requested. It is the VM's responsibility to handle user transactions, and include them into new blocks, mempool handling, and database handling.
+
+The consensus engine will just request the VM for new blocks whenever there is a signal (from VM) and according to other [congestion control mechanisms](https://github.com/ava-labs/avalanchego/blob/master/vms/proposervm/README.md), verify the block, gossip the block within the network for consensus, and request the VM to accept or reject the block.
```

Comment:
This paragraph does not fit under the interfaces header.

The content can use some cleaning up as well. Can we try:

ChainVMs are wrapped within the [ProposerVM]((https://github.com/ava-labs/avalanchego/blob/master/vms/proposervm/README.md)) to provide a soft leader. When a VM is ready to build a block, it will add a message to the `toEngine` channel that was passed in when the VM was initialized. The ProposerVM will intercept this message and wait for the time when it is that node's turn to propose a block on the chain to forward that message to the consensus engine. When the consensus engine receives that message, it will trigger a call to `BuildBlock()` on the VM, so that the VM can build a block to be issued into consensus.

Once the VM has built a block, the consensus engine will call verify on the block, and then issue it into consensus. The consensus engine actively drives consensus and whenever a block has been decided, the engine will call Accept/Reject on the block.